src/Compilers/CSharp/Test/Symbol/Symbols/Source/EventTests.cs

```diff
@@ -325,6 +325,397 @@ class C
+    public virtual event System.Action<object> E3 { add { System.Console.WriteLine(""Not called""); } remove {} }
+}";
+            var libComp = CreateCompilationWithMscorlib(text: libText, references: new[] { MscorlibRef, SystemCoreRef });
+            var libAssemblyRef = libComp.EmitToImageReference();
+            var source = @"
+class B : C
+{
+    public override event System.Action<dynamic> E1;
+    public override event System.Action<object> E2;
+    public override event System.Action<dynamic> E3;
+    public void RaiseE1(object s) => E1(s);
+    public void RaiseE2(string s) => E2(s);
+    public void RaiseE3(string s) => E3(s);
+}
+class A
+{
+    static void Main()
+    {
+        var b1 = new B();
+        b1.E1 += f => System.Console.WriteLine(f.Insert(0, ""Success1: ""));
+        b1.E2 += f => System.Console.WriteLine(""Success2"");
+        b1.E3 += f => System.Console.WriteLine(f.Insert(0, ""Success3: ""));
+        b1.RaiseE1(""Alice"");
+        b1.RaiseE2(""Bob"");
+        b1.RaiseE3(""Charlie"");
+
+        var b2 = new B();
+        b2.E1 += Print;
+        b2.E2 += Print;
+        b2.E3 += Print;
+        b2.RaiseE1(""Alice"");
+        b2.RaiseE2(""Bob"");
+        b2.RaiseE3(""Charlie"");
+    }
+
+    static void Print(object o) => System.Console.WriteLine(""Printed: "" + (System.String)o);
+}";
+            var expectedOutput =
+@"Success1: Alice
+Success2
+Success3: Charlie
+Printed: Alice
+Printed: Bob
+Printed: Charlie
+";
+            var compilationVerifier = CompileAndVerify(source: source, additionalRefs: new[] { MscorlibRef, SystemCoreRef, CSharpRef, libAssemblyRef },
+                                                    expectedOutput: expectedOutput);
+        }
+
+        [Fact, WorkItem(7845, "https://github.com/dotnet/roslyn/issues/7845")]
+        public void EventWithDynamicAttribute_OverriddenTypeDoesntLeak()
+        {
+            var libText = @"
+public class CL1
+{
+    public virtual event System.Action<dynamic> E1 { add {} remove {} }
+    public virtual event System.Action<dynamic> E2 { add {} remove {} }
+}";
+            var libComp = CreateCompilationWithMscorlib(text: libText, references: new[] { MscorlibRef, SystemCoreRef });
+            var libAssemblyRef = libComp.EmitToImageReference();
+
+            var source = @"
+public class CL2 : CL1
+{
+    public override event System.Action<object> E1;
+    public override event System.Action<object> E2 { add {} remove {} }
+
+    void M() {
+        E1 += (string s) => System.Console.WriteLine(s); // type won't fit
+        E2 += (string s) => System.Console.WriteLine(s); // type won't fit
+        E1(""Avoid warning"");
+    }
+}";
+            var comp = CreateCompilationWithMscorlib(text: source, references: new[] { libAssemblyRef }).VerifyDiagnostics(
+                // (8,9): error CS1661: Cannot convert lambda expression to delegate type 'Action<object>' because the parameter types do not match the delegate parameter types
```

Comment:
I am not sure what is the goal here. Do we want to verify that CL2.E1 and CL2.E2 still have type ```System.Action<object>```? If that is correct, let's test that directly rather than use diagnostics for this - create compilation, lookup symbols for the events and check their types.